crates/ra_hir/src/ids.rs

```diff
@@ -274,6 +274,17 @@ pub(crate) trait AstItemDef<N: AstNode>: salsa::InternKey + Clone {
         let loc = self.lookup_intern(db);
         loc.module
     }
+
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<N>) -> Option<Self> {
```

Comment:
Rather than adding a new trait method, let's move this to `from_source.rs` in the form of generic funciton:

```rust
fn from_source<Item: AstItemDef>(db: &(impl DefDatabase + AstDatabase), src: Source<N>) -> Option<Item>
```